### DISCUSSION TITLE
RabbitMQ sharding and Consistent hash exchange: not able to distribute messages across queues uniformly

### DISCUSSION THREAD

**Discussion by taju20910** [rabbitmq/rabbitmq-server#3608] [Other]
Version : 3.9.7

RabbitMQ sharding plugin or consistent hash exchange is not able to distribute messages uniformly across all queues. Messages always go to one common queue among the list of all bound queues.

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
We cannot suggest anything with the amount of information provided. Please help others help you and post some actual code that reproduces the behavior.

Both plugins [rely on routing key values to be reasonably unique and evenly distributed](https://github.com/rabbitmq/rabbitmq-server/tree/master/deps/rabbitmq_sharding#how-evenly-will-messages-be-distributed) or they won't be able to provide even routing target distribution. That's how hashing works: if the inputs are mostly the same, so are the outputs.

**Reply by taju20910** [reply]
Thanks Michael will look further and provide more details

**Reply by taju20910** [reply]
![queue](https://user-images.githubusercontent.com/32966851/138845132-99efd7bb-d727-412d-842c-5e44cfbc4b58.png)
![exchange](https://user-images.githubusercontent.com/32966851/138845157-67ddd937-82ad-4763-8e28-3f4b75cadfa6.png)
Please find one scenario for consistent hash with attached screenshot
I am creating 3 queues and binding them to hash_exchange with routing key 1234. when I am publishing messages in this exhange with routing key 1234 always messages go to one queue instead messages should distribute across three queues

@michaelklishin FYI

**Reply by michaelklishin** [reply]
We need executable code or scripts, not screenshots. If the routing key is identical for all messages, they all will be routed to the same queue. Again, this is how hashing works and this is documented at least for `rabbitmq_sharding`.

**Reply by michaelklishin** [reply]
If you want messages with the same routing key to be reasonably evenly distributed between all bound queues, see [`rabbitmq_random_exchange`](https://github.com/rabbitmq/rabbitmq-server/tree/master/deps/rabbitmq_random_exchange). It might need a new release for RabbitMQ 3.9 and it's hardly a perfect distribution mechanism but it does not care what the routing key is.

**Reply by taju20910** [reply]
Thanks @michaelklishin for all the help
